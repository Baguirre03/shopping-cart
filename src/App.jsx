import { Routes, Route } from "react-router-dom";
import Navbar from "./Nav";
import Home from "./Home";
import Shop from "./Shop";
import { useState } from "react";
import Cart from "./Cart";
import "./styles/app.css";

function App() {
  const [cart, setCart] = useState(0);
  const [cartItems, setCartItems] = useState([]);

  function minusCart(data) {
    if (!cartItems.filter((e) => e.title === data.title).length) {
      return;
    } else {
      setCart(cart - 1);
      handleSubtract(data);
    }
  }

  function addCart(data) {
    setCart(cart + 1);
    handleCartItems(data);
  }

  function handleCartItems(data) {
    let copy = [...cartItems];
    if (cartItems.filter((el) => el.title === data.title).length) {
      const index = copy.findIndex((el) => el.title === data.title);
      let item = copy[index];
      copy[index] = {
        ...item,
        count: (item.count += 1),
      };
    } else {
      copy.push({
        image: data.image,
        description: data.description,
        price: data.price,
        title: data.title,
        id: crypto.randomUUID(),
        count: 1,
      });
    }
    setCartItems(copy);
  }

  function handleSubtract(data) {
    let copy = [...cartItems];
    const index = copy.findIndex((el) => el.title === data.title);
    let item = copy[index];
    if (item.count > 1) {
      item = {
        ...item,
        count: (item.count -= 1),
      };
    } else {
      copy.splice(index, 1);
    }
    setCartItems(copy);
  }

  return (
    <>
      <Navbar cart={cart}></Navbar>
      <Routes>
        <Route path="/" element={<Home />}></Route>
        <Route
          path="shop"
          element={
            <Shop
              cart={cart}
              handleCart={handleCartItems}
              minusCart={minusCart}
              addCart={addCart}
            />
          }
        ></Route>
        <Route
          path="cart"
          element={
            <Cart
              cart={cart}
              cartItems={cartItems}
              handleCart={handleCartItems}
              minusCart={minusCart}
              addCart={addCart}
            />
          }
        ></Route>
      </Routes>
    </>
  );
}

export default App;
