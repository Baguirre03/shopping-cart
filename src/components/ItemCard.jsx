export default function ItemCard({ data, minusCart, addCart }) {
  console.log(data, "data");
  return (
    <div className="item">
      <div className="info">
        <img src={data.image} alt={data.description} />
        <h4>{data.title}</h4>
        <p>{data.description}</p>
      </div>
      <div>
        <button onClick={() => minusCart()}>-</button>
        <div>{data.price}</div>
        <button onClick={() => addCart(data)}>+</button>
      </div>
    </div>
  );
}
