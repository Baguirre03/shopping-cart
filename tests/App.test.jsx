import { getByDisplayValue, render } from "@testing-library/react"
import App from "../src/App"
import { expect } from "vitest"
import { BrowserRouter } from "react-router-dom"
import ItemCard from "../src/components/ItemCard"
import { screen } from "@testing-library/react"

describe('main page works', () => {
    it('loads the page', () => {
        const {container} = render(<BrowserRouter><App></App></BrowserRouter>)
        expect(container).toMatchSnapshot()
    })
})


describe('creates an item card correctly', () => {
    let cart = false
    let data = {
        image: null,
        description: 'test',
        title: 'title',
    }
    it('creates an item correctly', () => {
        render(<ItemCard data={data} cart={cart}></ItemCard>)
        expect(screen.getByRole('heading', {level: 4}).textContent).toMatch(/title/i)
    })
})