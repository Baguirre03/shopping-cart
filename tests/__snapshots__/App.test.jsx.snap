// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`main page works > loads the page 1`] = `
<div>
  <header>
    <h1>
      <a
        href="/"
      >
        <button>
          Super Cool Shop
        </button>
      </a>
    </h1>
    <nav>
      <a
        href="/"
      >
        <button>
          Home
        </button>
      </a>
      <a
        href="/shop"
      >
        <button>
          Shop
        </button>
      </a>
    </nav>
  </header>
  <div
    class="home-page"
  >
    <div
      class="center-content"
    >
      <span
        class="background"
      />
      <h2>
        Super Cool Shop
      </h2>
      <div>
        This is definitely a shop
      </div>
    </div>
  </div>
</div>
`;
